edge/pkg/edgehub/upgrade/image_prepull.go

```diff
@@ -0,0 +1,195 @@
+		return fmt.Errorf(errmsg)
+	}
+	err = json.Unmarshal(data, &prepullReq)
+	if err != nil {
+		errmsg = fmt.Sprintf("failed to unmarshal message content, err: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return fmt.Errorf(errmsg)
+	}
+	if nodeName != prepullReq.NodeName {
+		errmsg = fmt.Sprintf("request node name %s is not match with the edge node %s", prepullReq.NodeName, nodeName)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return fmt.Errorf(errmsg)
+	}
+
+	// todo check items such as disk according to req.checkitem config
+
+	// pull images
+	container, err := util.NewContainerRuntime(edgecoreconfig.Modules.Edged.TailoredKubeletConfig.ContainerRuntimeEndpoint, edgecoreconfig.Modules.Edged.TailoredKubeletConfig.CgroupDriver)
+	if err != nil {
+		errmsg = fmt.Sprintf("failed to new container runtime: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return fmt.Errorf(errmsg)
+	}
+
+	go prepullImages(nodeName, jobName, prepullReq, container)
+	return nil
+}
+
+func sendResponseMessage(nodeName, jobName, pullMsg string, state v1alpha1.PrepullState) {
+	resp := commontypes.ImagePrepullJobResponse{
+		NodeName: nodeName,
+		State:    state,
+		Message:  pullMsg,
+	}
+	msg := model.NewMessage("").SetRoute(modules.EdgeHubModuleName, modules.HubGroup).
+		SetResourceOperation(fmt.Sprintf("node/%s/prepull/%s", nodeName, jobName), "prepull").FillBody(resp)
+	beehiveContext.Send(modules.EdgeHubModuleName, *msg)
+}
+
+func prepullImages(nodeName, jobName string, prepullReq commontypes.ImagePrepullJobRequest, container util.ContainerRuntime) {
+	authConfig, err := makeAuthConfig(prepullReq.Secret)
+	if err != nil {
+		errmsg := fmt.Sprintf("failed to get prepull secret, err: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return
+	}
+
+	state := v1alpha1.PrepullSuccessful
+	var respStatus []ImageStatusEachNode
+	for _, image := range prepullReq.Images {
+		prepullStatus := ImageStatusEachNode{
+			Image: image,
+		}
+		for i := 0; i < int(prepullReq.RetryTimes); i++ {
+			err = container.PullImage(image, authConfig, nil)
+			if err == nil {
+				break
+			}
+		}
+		if err != nil {
+			klog.Errorf("pull image %s failed, err: %v", image, err)
+			state = v1alpha1.PrepullFailed
+			prepullStatus.State = v1alpha1.PrepullFailed
+			prepullStatus.Reason = err.Error()
+		} else {
+			klog.Infof("pull image %s successfully!", image)
+			prepullStatus.State = v1alpha1.PrepullSuccessful
+		}
+		respStatus = append(respStatus, prepullStatus)
+	}
+
+	respData, err := json.Marshal(respStatus)
+	if err != nil {
+		errmsg := fmt.Sprintf("marshal response status failed, err: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return
+	}
+	sendResponseMessage(nodeName, jobName, string(respData), state)
+}
+
+func parseJobName(resource string) (string, error) {
+	var jobName string
```

Comment:
These variables don't seem to matter, right?

Reply:
ditto.